### PR TITLE
cranelift-isle: Add "partial" flag for constructors

### PR DESCRIPTION
**Pull request by jameysharp**
I speculate that this PR may have a performance impact, which I want to measure before merging, but I'd appreciate review on the implementation. Also I wanted to get the CI run going overnight.

Instead of tying fallibility of constructors to whether they're either internal or pure, this commit assumes all constructors are infallible unless tagged otherwise with a "partial" flag.

Internal constructors without the "partial" flag are not allowed to use constructors which have the "partial" flag on the right-hand side of any rules, because they have no way to report last-minute match failures.

Multi-constructors should never be "partial"; they report match failures with an empty iterator instead. In turn this means you can't use partial constructors on the right-hand side of internal multi-constructor rules. However, you can use the same constructors on the left-hand side with `if` or `if-let` instead.

In many cases, ISLE can already trivially prove that an internal constructor always returns `Some`. With this commit, those cases are laregly unchanged, except for removing all the `Option`s and `Some`s from the generated code for those terms.

However, for internal non-partial constructors where ISLE could not prove that, it now emits an `unreachable!` panic as the last-resort, instead of returning `None` like it used to do. Among the existing backends, here's how many constructors have these panic cases:

- x64: 14% (53/374)
- aarch64: 15% (41/277)
- riscv64: 23% (26/114)
- s390x: 47% (268/567)

It's often possible to rewrite rules so that ISLE can tell the panic can never be hit. Just ensure that there's a lowest-priority rule which has no constraints on the left-hand side.

But in many of these constructors, it's difficult to statically prove the unhandled cases are unreachable because that's only down to knowledge about how they're called or other preconditions.

So this commit does not try to enforce that all terms have a last-resort fallback rule.

### CODE REVIEW

**Review by cfallin — APPROVED** [review]
Looks great, thanks! Just pair-reviewed this with Jamey; nothing too surprising going on, and it's a nice generalization of the sorts of constructors we support today.

**Comment by jameysharp** [comment]
On a suggestion from Chris, I ran `wasmtime compile` on the Spidermonkey benchmark from Sightglass, both with and without this PR, and compared the compiled .cwasm files. They were byte-for-byte identical. In conjunction with the fact that none of the precise-output filetests failed on any backend, I have pretty good confidence that this didn't change the results of instruction selection.

I had thought this PR might marginally improve performance in the instruction selection phase of compiling, due to fewer checks for whether constructors returned `Option::Some`.

When measuring time spent in `wasmtime compile` on Spidermonkey with DHAT, this PR reduces instructions retired by 0.06%, at the cost of 0.0007% more bytes read and 0.0001% more bytes written. Somehow this PR increased total heap allocations by 8 bytes.

Sightglass/perf agrees with DHAT that instructions retired during compilation are improved by a statistically significant amount but with such a small effect size that it reports this PR is "1.00x faster" on the bz2, pulldown-cmark, and spidermonkey benchmarks.

Sightglass' cpu-cycles measure and Hyperfine's wall-clock time both report that this PR is slightly slower when compiling Spidermonkey: the baseline version is "1.00x to 1.01x faster" according to Sightglass and "1.00 ± 0.02 times faster" according to Hyperfine. On the smaller benchmarks (bz2 and pulldown-cmark), Sightglass found no significant difference in cpu-cycles.

In short, this PR has almost no effect on performance by any measure.
